pkg/restore/restore.go

```diff
@@ -634,20 +636,73 @@ func (ctx *context) restoreResource(resource, namespace, resourcePath string) (a
 		}
 
 		// necessary because we may have remapped the namespace
-		obj.SetNamespace(namespace)
+		// if the namespace is blank, don't create the key
+		if namespace != "" {
+			obj.SetNamespace(namespace)
+		}
 
 		// add an ark-restore label to each resource for easy ID
 		addLabel(obj, api.RestoreLabelKey, ctx.restore.Name)
 
 		ctx.infof("Restoring %s: %v", obj.GroupVersionKind().Kind, obj.GetName())
-		_, err = resourceClient.Create(obj)
-		if apierrors.IsAlreadyExists(err) {
-			addToResult(&warnings, namespace, err)
-			continue
+		// Use a special error since we need to retain that information through a couple checks.
+		_, restoreErr := resourceClient.Create(obj)
+		if apierrors.IsAlreadyExists(restoreErr) {
+			fromCluster, err := resourceClient.Get(obj.GetName(), metav1.GetOptions{})
+			if err != nil {
+				addToResult(&warnings, namespace, err)
+				continue
+			}
+			// Remove insubstantial metadata
+			fromCluster, err = resetMetadataAndStatus(fromCluster)
+			if err != nil {
+				addToResult(&warnings, namespace, err)
+				continue
+			}
+			if obj.GroupVersionKind().Kind == "ServiceAccount" {
+				desired, err := mergeSecrets(fromCluster, obj)
+				if err != nil {
+					addToResult(&warnings, namespace, err)
+					continue
+				}
+				desiredBytes, err := json.Marshal(desired)
+				if err != nil {
+					addToResult(&warnings, namespace, err)
+					continue
+				}
+				fromClusterBytes, err := json.Marshal(obj)
+				if err != nil {
+					addToResult(&warnings, namespace, err)
+					continue
+				}
+				patchBytes, err := jsonmergepatch.CreateThreeWayJSONMergePatch(fromClusterBytes, desiredBytes, nil)
```

Comment:
Why 3-way?

Reply:
I was trying to use the higher level APIs; I can drop down to using the evanphx lib directly, though.